Improve reflected attribute tests (html/dom/reflection.js)

tbroyer opened the pull request:
Fixes #44315

tbroyer committed:
Fix reflection tests wrt valueOf

Without a toString:null in test values that are objects
with a valueOf method, the DOM attribute would actually
be set to "[object Object]" without ever calling the
valueOf method (effectively turning those tests into
behaving exactly the same as the {"test": 6} ones).

Most existing tests were passing because the implemented
algorithms would actually compute a WRONG expected value.
The only test that doesn't compute the expected value
("double") won't actually run because all reflected IDL
attributes of type "double" have custom getters.

Fixes #44315

tbroyer committed:
Fix reflection tests wrt -0 for "double" attributes

Setting -0 to the IDL attribute will set the DOM
attribute to "0" (per ECMAScript's ToString, from the
rules for the the best representation of a number as
a floating point number), and the rules for parsing
flowting-point number values won't ever turn that "0"
into a -0 value.

This test value is never actually being used though,
as all the reflected attributes of type "double" have
custom getters.

tbroyer committed:
Implement "limited double" in reflection tests

tbroyer committed:
Implement the reflect algorithm for doubles

This should make it easier to extend tests with more values.

tbroyer commented:
Various improvements, each in its own commit. Let me know if you'd prefer separate PRs.

sideshowbarker commented on html/dom/reflection.js:
```diff
@@ -522,24 +600,68 @@ ReflectionTests.typeMap = {
+        "defaultVal": 0.0,
+        "domTests": [minInt - 1, minInt, -36, -1, 0, 1, maxInt,
+            maxInt + 1, maxUnsigned, maxUnsigned + 1, "",
+            "\u00097", "\u000B7", "\u000C7", "\u00207", "\u00A07", "\uFEFF7",
```
In working on a patch for https://webkit.org/b/268542, I find it’d be useful to test that a U+000B isn’t handled as allowed whitespace even when the U+000B is other than the first code point in the value.
```suggestion
            "\u00097", "\u000B7", "\u000C7", "\u00207", "\u00A07", "\uFEFF7",
            "\t\u000B7", "\n\u000B7","\f\u000B7", "\r\u000B7", "\x20\u000B7",
```

tbroyer replied:
Ah interesting. This should probably be tested for all numeric values (all the longs) as well.

How should I amend the PR? Amend the appropriate commit (rebase with fixups) or pushing new commits? What are the rules here? (I don't think I saw this discussed in the docs)

sideshowbarker replied:
> How should I amend the PR? Amend the appropriate commit (rebase with fixups) or pushing new commits? What are the rules here?

I don’t think there’s any strict policy. So it’s fine to amend and force-push, if that’s your normal workflow elsewhere. I guess it’s also fine to push new commits, if you want to be able to roll back any changes or whatever.

As far as the reviewer workflow of reviewing locally in the shell from the checked-out PR branch, I guess there it’s easier to review when changes are amended and force-pushed rather than just pushed as new commits.

As far as the reviewer workflow of reviewing from the GitHub UI, I’m not sure how most other people review PRs with the GitHub UI — but personally, I never review from the **Commits** tab; instead I always review from the **Files changed** tab. So for the GitHub UI case for me at least, it doesn’t really matter whether somebody has pushed new commits instead of amending and force-pushing. I suspect that probably holds true for most other people as well.

tbroyer replied:
Added those tests to all numeric types.

tbroyer commented on html/dom/reflection.js:
```diff
@@ -521,23 +601,23 @@ ReflectionTests.typeMap = {
-        "idlDomExpected": ["-10000000000", "-1", "0", "0", "1", "10000000000"],
-        "idlIdlExpected": [ -10000000000,   -1,   0,   0,   1,   10000000000]
+        "idlDomExpected": function (val) {
+            return Number(val).toFixed();
```
This is actually wrong, it should be just `Number(val)`, which would be the default behavior when idlDomExpected is absent.
This wouldn't change the behavior for the test values, but could be wrong should new values be tested.

https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#best-representation-of-the-number-as-a-floating-point-number

tbroyer replied:
I amended the commits and force-pushed. Hope this is OK for you.

tbroyer committed:
Add a few more numeric values to reflection tests

Specifically "+" and "-" to exercise all possible code
paths of failures.

tbroyer committed:
Add exponential notation to reflection tests of double values

tbroyer committed:
Add a few more reflection tests around whitespace for numeric types

Specifically leading non-ASCII whitespace following leading
ASCII whitespace (should be rejected), and trailing non-ASCII
whitespace (should be ignored)

tbroyer commented:
I ran the tests with coverage (a copy, in a different project that reuses them), which led me to add a few tests (including exponential notation) and revealed a bug in the `parseFloat` function implementing the HTML spec rules for parsing floating-point number values. Once the test harness was fixed, it didn't reveal any more bugs in browsers (phew!)

PR updated with the new tests and fixed harness.

zcorpan commented on html/dom/reflection.js:
```diff
@@ -87,6 +87,88 @@ ReflectionTests.parseInt = function(input) {
+        }
+    }
+    // Conversion:
+    // TODO: round to the closest even value, and return an error if it rounds to infinity
```
I believe using JS numbers you get rounding for free, except for Infinity and -Infinity which you can check for with `===`.

tbroyer replied:
Ack. I used Number.isFinite: the algorithm won't produce NaN, so I assume it's ok to test for infinities or NaN when all we want to detect are infinities.

zcorpan commented on html/dom/reflection.js:
```diff
@@ -501,45 +583,82 @@ ReflectionTests.typeMap = {
+            "\t\u000B7", "\n\u000B7","\f\u000B7", "\r\u000B7", "\x20\u000B7", "7\u000B",
             " " + binaryString + " foo ", undefined, 1.5, "5%", "+100", ".5", true, false,
+            "1.", "1e2", "1e+2", "1e-2", "1E2", "1E+2", "1E-2", "1.e2", "1.0e2",
+            "1. 1", "1 .1", "1. e2", "1 .e2", "1 e2", "1e 2", "1e -2", "1e- 2",
```
Maybe add strings that parse to Infinity and -Infinity, e.g. `"1.8e308"` and `"-1.8e308"`

zcorpan reviewed (changes requested):
Generally looks good, thanks! `parseFloat` I think matches the spec except for infinities. (Maybe the spec should have this structure instead of goto-style.)
